### PR TITLE
feat: implement lazy set and delete of legacy nodes

### PR DESCRIPTION
**Pull request by cool-develope**
- create the different prefix for the new format node key to avoid the complex when removing nodes
- delete all legacy versions at once in `DeleteVersionsTo` to avoid duplicated codes
- this PR will not include the updating of legacy node to new format

### CODE REVIEW

File: nodedb.go [hunk 1]
```diff
@@ -41,11 +41,11 @@ const (
 
 var (
 	// All node keys are prefixed with the byte 'n'. This ensures no collision is
```
Comment (yihuang) [inline nodedb.go:43]:
```suggestion
	// All new node keys are prefixed with the byte 's'. This ensures no collision is
```

File: nodedb.go [hunk 2]
```diff
@@ -122,17 +133,32 @@ func (ndb *nodeDB) GetNode(nk []byte) (*Node, error) {
 	ndb.opts.Stat.IncCacheMissCnt()
 
 	// Doesn't exist, load.
-	buf, err := ndb.db.Get(ndb.nodeKey(nk))
```
Comment (yihuang) [inline nodedb.go:125]: `nk` is the node key but without prefix?

File: node.go [hunk 3]
```diff
@@ -171,6 +168,69 @@ func MakeNode(nk []byte, buf []byte) (*Node, error) {
...
+		if cause != nil {
+			return nil, fmt.Errorf("deocding node.leftHash, %w", cause)
+		}
+		buf = buf[n:]
+
+		rightHash, _, cause := encoding.DecodeBytes(buf)
+		if cause != nil {
+			return nil, fmt.Errorf("decoding node.rightHash, %w", cause)
+		}
+		node.leftNodeKey = leftHash
```
Comment (yihuang) [inline node.go:228]: it should be the node key with prefix here?

**Review by yihuang — CHANGES_REQUESTED** [review]
- should we store the prefixed node key in the `leftChildKey` and `rightChildKey`?
- can you add some unit tests for the mixed nodes scenario.

File: nodedb.go (hunk 2 above)
Comment (cool-develope) [reply nodedb.go:125]: yes

**Comment by cool-develope** [comment]
> * should we store the prefixed node key in the `leftChildKey` and `rightChildKey`?

Here, the concern is it will increase the node encoding size

> * can you add some unit tests for the mixed nodes scenario.

I added the legacy db generator commands and unit-tests

**Comment by cool-develope** [comment]
the lazy update is a little tricky, the same problem with `Import/Export`, how to track the current nonce with the specific version when updating the legacy nodes?
@yihuang have you any ideas?

**Comment by yihuang** [comment]
> the lazy update is a little tricky, the same problem with `Import/Export`, how to track the current nonce with the specific version when updating the legacy nodes? @yihuang have you any ideas?

WDYM? we just need to use a increasing nonce?

**Pull request merged**